app/Http/Controllers/SupplyController.php

```diff
@@ -0,0 +1,269 @@
+            $response->headers->set("Access-Control-Allow-Origin", $request->headers->get("Origin"));
+            $response->headers->set("Access-Control-Allow-Credentials", "true");
+            $response->headers->set("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
+        }
+
+        if ($request->getRealMethod() == 'GET') {
+            $data = $request->getQueryString();
+        } elseif ($request->getRealMethod() == 'POST') {
+            $data = $request->getContent();
+        } elseif ($request->getRealMethod() == 'OPTIONS') {
+            $response->setStatusCode(204);
+            $response->headers->set('Access-Control-Max-Age', 1728000);
+            return $response;
+        }
+
+        $decoded = Utils::decodeZones($data);
+//         print_r($decoded);exit;
+        $zones = $decoded['zones'];
+
+        $tid = Utils::attachTrackingCookie(config('app.adserver_secret'), $request, $response, "", new \DateTime());
+
+        // use adselect here
+        $context = Utils::getImpressionContext($request, $data);
+
+        $impressionId = $decoded['page']['iid'];
+        if ($impressionId) {
+            $aduser_endpoint = config('app.aduser_endpoint');
+            if ($aduser_endpoint) {
+                $userdata = (array)json_decode(file_get_contents("{$aduser_endpoint}/get/{$impressionId}"), true);
+            } else {
+                $userdata = [];
+            }
+        }
+
+        $keywords = array_merge($context, $userdata);
+
+        $banners = BannerFinder::getBestBanners($zones, $keywords);
+
+        foreach ($banners as &$banner) {
+            if ($banner) {
+                $banner['pay_to'] = Esc::normalizeAddress(config('app.adshares_address'));
+            }
+        }
+
+        $response->setContent(json_encode($banners, JSON_PRETTY_PRINT));
+        return $response;
+    }
+
+    public function findScript(Request $request)
+    {
+        $params = [json_encode($request->getSchemeAndHttpHost()), json_encode(config('app.aduser_endpoint'))];
+
+        $jsPath = config('app.env') == 'production' ? resource_path('assets/js/tmp-copy/find.x.js') : resource_path('assets/js/tmp-copy/find.min.js');
+
+        $response = new StreamedResponse();
+        $response->setCallback(function () use ($jsPath, $request, $params) {
+            echo str_replace([
+                "'{{ ORIGIN }}'",
+                "'{{ ADUSER }}'",
+            ], $params, file_get_contents($jsPath));
+        });
+
+        $response->headers->set('Content-Type', 'text/javascript');
+
+        $response->setCache(array(
+            'etag' => md5(md5_file($jsPath) . implode(':', $params)),
+            'last_modified' => new \DateTime('@' . filemtime($jsPath)),
+            'max_age' => 3600 * 24 * 30,
+            's_maxage' => 3600 * 24 * 30,
+            'private' => false,
+            'public' => true
+        ));
+
+        if (! $response->isNotModified($request)) {
+            // TODO: ask Jacek
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1135)